src/modules/ipc.cpp

```diff
@@ -103,6 +109,15 @@ namespace modules {
     return m_builder->flush();
   }
 
+  string ipc_module::get_format() const {
+      if( m_current_hook != -1 && (size_t)m_current_hook < m_hooks.size()){
+        return "format-" + to_string(m_current_hook);
+      }
+      else{
+        throw module_error("Initial hook out of bounds: '" + to_string(m_initial + 1) +
```

Comment:
This crashes the module if `initial` is not set, or set to 0.

We guarantee that `-1 <= m_current_hook < m_hooks.size()`. So, we can simply return `DEFAULT_FORMAT` if `has_hook()` returns false and otherwise return the constructed format name.

Reply:
I think we need to keep some sort of mapping (like unordered_map) for hook-x to the internal index we store.

Reply:
A simple fallback could also be implemented by adding the `format-N` using `m_formatter->add_optional` and then in `get_format` first checking `m_formatter->has_format` and if it doesn't exist, fall back to the default format.

Reply:
> A simple fallback could also be implemented by adding the `format-N` using `m_formatter->add_optional` and then in `get_format` first checking `m_formatter->has_format` and if it doesn't exist, fall back to the default format.

I tried this, add_optional is not adding the format to the list, in addition , I noticed that this config
```
hook-0 = echo foobar
hook-1 = date +%s
;hook-2 = whoami
hook-3 = pwd
format= <label>
format-0-foreground= ${color.foreground}
format-1-foreground= ${color.foreground}
;format-2-foreground= ${color.foreground}

format-0-background= ${color.shade1}
format-1-background= ${color.shade2}
;format-2-background= ${color.shade3}
initial = 1
click-left = "#demo.hook.0"
click-right = "#demo.hook.1"
;double-click-left = "#demo.hook.2"
```
doesnt detect hook-3, because hook-2 stuff is commented. ( is this a rule in configs where you define hook-x in increasing order from 0 to n).  Im trying to figure out why m_conf.has(m_modname, name) is failing.

Reply:
> I tried this, add_optional is not adding the format to the list,

Yeah, the disadvantage of that is that it only detects a format if the key `format-N` is defined. If you only define `format-N-foreground`, it doesn't see it.

And yes, it's a rule that lists like this have to be complete. So for `hook-3` to be loaded, `hook-0` up to `hook-2` have to exist as well.

Reply:
> Yeah, the disadvantage of that is that it only detects a format if the key `format-N` is defined. If you only define `format-N-foreground`, it doesn't see it.

If I add format-N in the config, there is no output from the IPC module. In this case, neither ```add()``` nor ```add_optional()``` works. If I search for format-n in  ipc::constructor with ```has_format ("format-n")```, the result turns out negative.

Reply:
Then there is probably something else going wrong. If the setting `format-0` exists, both `add("format-0")` and `add_optional("format-0")` will load the format and `has_format("format-0")` will return true.

Reply:
> Then there is probably something else going wrong. If the setting `format-0` exists, both `add("format-0")` and `add_optional("format-0")` will load the format and `has_format("format-0")` will return true.

polybar|error: Failed to get contents for "module/demo" (err: Format "format-0" has not been added)
, throw undefined_format is being thrown. This is what happens when I define format-0 in config, ```has_format()``` fails, 
![image](https://user-images.githubusercontent.com/59736487/188345293-1fd699fa-7ff8-4d17-ac1b-21eb610e6a66.png).

Reply:
How are you adding the formats to the formatter?

> My doubt is, why does m_formats only have "format", shouldnt it have the other keywords of module/demo

`m_formats` only contains the formats added with `add` or `add_optional`.

`m_conf` is the thing that contains all config options

Reply:
> How are you adding the formats to the formatter?
In this case, I commented any addition of format. But even if I add, with either add or add_optional (after defining format-n in config), the module is not visible. I guess this error was due to the fact that I wasn't adding the format.